### PR TITLE
Extension - Do not show Signature popup in case not enough balance to cover gas fee

### PR DESCRIPTION
**Pull request by Thiendekaco**

### CODE REVIEW

**Comment by saltict** [comment]
🚀 Deployed on https://pr-3382--sw-web-runner.netlify.app

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 1]
```diff
@@ -164,9 +164,11 @@ export default class TransactionService {
     // Fill transaction default info
     const transaction = this.fillTransactionDefaultInfo(inputTransaction);
 
-    // Add Transaction
-    transactions[transaction.id] = transaction;
-    this.transactionSubject.next({ ...transactions });
+    if (!transaction.errors || transaction.errors.length === 0) {
```
Comment (S2kael) [inline packages/extension-base/src/services/transaction-service/index.ts:167]: Why have this logic?

File: packages/extension-base/src/core/logic-validation/request.ts [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+
+import { assert, isString } from '@polkadot/util';
+import { isEthereumAddress } from '@polkadot/util-crypto';
+
+export type ValidateStepFunction = (koni: KoniState, url: string, payload: PayloadValidated, topic?: string) => Promise<PayloadValidated>
+
+export interface PayloadValidated {
+  networkKey?: string,
+  address: string,
+  pair?: KeyringPair,
```
Comment (S2kael) [inline packages/extension-base/src/core/logic-validation/request.ts:28]: Make it required

**Review by S2kael — CHANGES_REQUESTED** [review]

**Review by S2kael — APPROVED** [review]

File: packages/extension-base/src/core/logic-validation/request.ts [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+  try {
+    transaction.nonce = await web3.eth.getTransactionCount(fromAddress);
+  } catch (e) {
+    console.error(e);
+    errors.push(new TransactionError(BasicTxErrorType.INTERNAL_ERROR, handleErrorMessage(e as Error)));
+  }
+
+  const hasError = (errors && errors.length > 0) || !networkKey;
+  const hashPayload = hasError ? '' : koni.transactionService.generateHashPayload(networkKey, transaction);
+  const isToContract = !hasError && await isContractAddress(transaction.to || '', evmApi);
```
Comment (S2kael) [inline packages/extension-base/src/core/logic-validation/request.ts:266]: Is only network error will make this line stuck?

**Review by S2kael — CHANGES_REQUESTED** [review]

File: packages/extension-base/src/background/KoniTypes.ts [hunk 4]
```diff
@@ -1334,7 +1337,8 @@
   switchNetworkRequest: [ConfirmationsQueueItem<SwitchNetworkRequest>, ConfirmationResult<boolean>],
   evmSignatureRequest: [ConfirmationsQueueItem<EvmSignatureRequest>, ConfirmationResult<string>],
   evmSendTransactionRequest: [ConfirmationsQueueItem<EvmSendTransactionRequest>, ConfirmationResult<string>]
-  evmWatchTransactionRequest: [ConfirmationsQueueItem<EvmWatchTransactionRequest>, ConfirmationResult<string>]
+  evmWatchTransactionRequest: [ConfirmationsQueueItem<EvmWatchTransactionRequest>, ConfirmationResult<string>],
+  errorConnectNetwork: [ConfirmationsQueueItem<PayloadValidated>, ConfirmationResult<null>]
```
Comment (S2kael) [inline packages/extension-base/src/background/KoniTypes.ts:1341]: Create new type for this request, make it detailed

File: packages/extension-base/src/core/logic-validation/request.ts [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+      estimateGas,
+      hashPayload,
+      isToContract,
+      parseData,
+      canSign: !hasError
+    }
+  };
+}
+
+export async function validationEvmSignMessageMiddleware (koni: KoniState, url: string, payload_: PayloadValidated, validationMiddlewareSteps: ValidateStepFunction[]): Promise<PayloadValidated> {
```
Comment (S2kael) [inline packages/extension-base/src/core/logic-validation/request.ts:320]: Remove `validationMiddlewareSteps`

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 6]
```diff
@@ -164,9 +164,11 @@ export default class TransactionService {
...
+      // Add Transaction
```
Comment (S2kael) [inline packages/extension-base/src/services/transaction-service/index.ts:168]: Move this to `fillTransactionDefaultInfo`

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 7]
```diff
@@ -839,7 +841,7 @@ export default class TransactionService {
     const payload = (transaction as EvmSendTransactionRequest);
     const evmApi = this.state.chainService.getEvmApi(chain);
     const chainInfo = this.state.chainService.getChainInfoByKey(chain);
-
+    const hasError = !!(payload.errors && payload.errors.length > 0);
```
Comment (S2kael) [inline packages/extension-base/src/services/transaction-service/index.ts:844]: Can remove this

File: packages/extension-koni-ui/src/Popup/Confirmations/index.tsx [hunk 8]
```diff
@@ -33,9 +33,12 @@ const titleMap: Record<ConfirmationType, string> = {
   signingRequest: detectTranslate('Signature request'),
   switchNetworkRequest: detectTranslate('Add network request'),
   connectWCRequest: detectTranslate('WalletConnect'),
-  notSupportWCRequest: detectTranslate('WalletConnect')
+  notSupportWCRequest: detectTranslate('WalletConnect'),
+  errorConnectNetwork: detectTranslate('Request Error')
 } as Record<ConfirmationType, string>;
 
+const ERROR_CONFIRMATION_TYPE = ['errorConnectNetwork'];
```
Comment (S2kael) [inline packages/extension-koni-ui/src/Popup/Confirmations/index.tsx:40]: Declare more specifically, type cast

File: packages/extension-koni-ui/src/Popup/Confirmations/parts/Sign/Evm.tsx [hunk 9]
```diff
@@ -71,7 +74,7 @@ const Component: React.FC<Props> = (props: Props) => {
   const isLedger = useMemo(() => signMode === AccountSignMode.LEGACY_LEDGER || signMode === AccountSignMode.GENERIC_LEDGER, [signMode]);
   const [showQuoteExpired, setShowQuoteExpired] = useState<boolean>(false);
   const isMessage = isEvmMessage(payload);
-
+  const isErrorTransaction = useMemo(() => errors && errors.length > 0, [errors]);
   const [loading, setLoading] = useState(false);
 
   const { error: ledgerError,
```
Comment (S2kael) [inline packages/extension-koni-ui/src/Popup/Confirmations/parts/Sign/Evm.tsx:80]: Add logic to prevent ledger connections

**Review by S2kael — CHANGES_REQUESTED** [review]

File: packages/extension-base/src/background/KoniTypes.ts [hunk 10]
```diff
@@ -1328,13 +1330,20 @@
   contractError: boolean;
 }
 
+export interface ErrorNetworkConnection {
+  networkKey: string,
+  address: string,
+  errors: Error[]
```
Comment (S2kael) [inline packages/extension-base/src/background/KoniTypes.ts:1336]: Does data transfer to UI success?

File: packages/extension-base/src/services/transaction-service/index.ts [hunk 11]
```diff
@@ -141,9 +141,9 @@ export default class TransactionService {
     return this.transactionSubject;
   }
 
-  private fillTransactionDefaultInfo (transaction: SWTransactionInput): SWTransaction {
+  private fillTransactionDefaultInfo (transaction: SWTransactionInput, useTransactionId?: boolean): SWTransaction {
     const isInternal = !transaction.url;
-    const transactionId = getTransactionId(transaction.chainType, transaction.chain, isInternal, isWalletConnectRequest(transaction.id));
+    const transactionId = useTransactionId ? transaction.id : getTransactionId(transaction.chainType, transaction.chain, isInternal, isWalletConnectRequest(transaction.id));
```
Comment (S2kael) [inline packages/extension-base/src/services/transaction-service/index.ts:146]: Must have `useTransactionId`?

**Review by S2kael — APPROVED** [review]

**Review by S2kael — APPROVED** [review]

**Review by saltict — APPROVED** [review]

**Pull request merged**